src/uct/tcp/tcp_cm.c

```diff
@@ -8,100 +8,336 @@
+                   (old_conn_state == UCT_TCP_EP_CONN_STATE_ACCEPTING));
+        if (old_conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK) {
+            uct_tcp_iface_outstanding_dec(iface);
+        }
+        break;
+    case UCT_TCP_EP_CONN_STATE_CLOSED:
+        ucs_assert(old_conn_state != UCT_TCP_EP_CONN_STATE_CLOSED);
+        if ((old_conn_state == UCT_TCP_EP_CONN_STATE_CONNECTING) ||
+            (old_conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK)) {
+            uct_tcp_iface_outstanding_dec(iface);
+        }
+        break;
+    default:
+        ucs_assert(ep->conn_state == UCT_TCP_EP_CONN_STATE_ACCEPTING);
+        /* Since ep::peer_addr is 0'ed and client's <address:port>
+         * has already been logged, no need to log here */
+        return;
+    }
+
+    ucs_debug("tcp_ep %p: %s -> %s for the [%s]<->[%s] connection",
+              ep, uct_tcp_ep_cm_state[old_conn_state].name,
+              uct_tcp_ep_cm_state[ep->conn_state].name,
+              ucs_sockaddr_str((const struct sockaddr*)&iface->config.ifaddr,
+                               str_local_addr, UCS_SOCKADDR_STRING_LEN),
+              ucs_sockaddr_str((const struct sockaddr*)&ep->peer_addr,
+                               str_remote_addr, UCS_SOCKADDR_STRING_LEN));
+}
+
+static void uct_tcp_cm_io_err_handler_cb(void *arg, int errno)
+{
+    uct_tcp_ep_t *ep = (uct_tcp_ep_t*)arg;
+
+    /* check whether this is possible somaxconn exceeded reason or not */    
+    if (((ep->conn_state == UCT_TCP_EP_CONN_STATE_CONNECTING) ||
+         (ep->conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK)) &&
+        ((errno == ECONNRESET) || (errno == ECONNREFUSED))) {
+        ucs_error("try to increase \"net.core.somaxconn\" on the remote node");
+    }
+}
+
+static void uct_tcp_cm_trace_conn_pkt(const uct_tcp_ep_t *ep, const char *msg,
+                                      const struct sockaddr_in *peer_addr)
+{
+    char str_addr[UCS_SOCKADDR_STRING_LEN];
+
+    ucs_debug("tcp_ep %p: %s %s", ep, msg,
+              ucs_sockaddr_str((const struct sockaddr*)peer_addr,
+                               str_addr, UCS_SOCKADDR_STRING_LEN));
+}
+
+static ucs_status_t
+uct_tcp_cm_conn_pkt_check_event(const uct_tcp_ep_t *ep,
+                                uct_tcp_cm_conn_event_t expected_event,
+                                uct_tcp_cm_conn_event_t actual_event)
+{
+    char str_addr[UCS_SOCKADDR_STRING_LEN];
+
+    if (expected_event != actual_event) {
+        ucs_error("tcp_ep %p: received wrong CM event (expected: %u, "
+                  "actual: %u) from the peer with iface listener "
+                  "address: %s", ep, expected_event, actual_event,
+                  ucs_sockaddr_str((const struct sockaddr*)&ep->peer_addr,
+                                   str_addr, UCS_SOCKADDR_STRING_LEN));
+        return UCS_ERR_INVALID_PARAM;
+    }
+    return UCS_OK;
+}
+
+static void uct_tcp_cm_handle_maxconn_exceed(ucs_status_t status)
```

Comment:
rename to uct_tcp_cm_handle_error
we don't really know it's maxconn when we call this function

Reply:
I forgot to remove this part of the code. fixed the comment in the latest commit